sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -0,0 +1,230 @@
+from kfp.v2 import dsl
+from kfp.v2.compiler import importer_node
+from kfp.v2.dsl import type_utils
+from kfp.v2.proto import pipeline_spec_pb2
+
+from google.protobuf import json_format
+
+
+class Compiler(object):
+  """Experimental DSL compiler that targets the PipelineSpec IR.
+
+  It compiles pipeline function into PipelineSpec json string.
+  PipelineSpec is the IR protobuf message that defines a pipeline:
+  https://github.com/kubeflow/pipelines/blob/237795539f7b85bac77435e2464367226ee19391/api/v2alpha1/pipeline_spec.proto#L8
+  In this initial implementation, we only support components authored through
+  Component yaml spec. And we don't support advanced features like conditions,
+  static and dynamic loops, etc.
+
+  Example:
+    How to use the compiler to construct pipeline_spec json:
+
+      @dsl.pipeline(
+        name='name',
+        description='description'
+      )
+      def my_pipeline(a: int = 1, b: str = "default value"):
+        ...
+
+      kfp.v2.compiler.Compiler().compile(my_pipeline, 'path/to/pipeline.json')
+  """
+
+  def _create_pipeline_spec(
+      self,
+      args: List[dsl.PipelineParam],
+      pipeline: dsl.Pipeline,
+  ) -> pipeline_spec_pb2.PipelineSpec:
+    """Creates the pipeline spec object.
+
+    Args:
+      args: The list of pipeline arguments.
+      pipeline: The instantiated pipeline object.
+
+    Returns:
+      A PipelineSpec proto representing the compiled pipeline.
+
+    Raises:
+      NotImplementedError if the argument is of unsupported types.
+    """
+    pipeline_spec = pipeline_spec_pb2.PipelineSpec()
+    if not pipeline.name:
+      raise ValueError('Pipeline name is required.')
+    pipeline_spec.pipeline_info.name = pipeline.name
+    pipeline_spec.sdk_version = 'kfp-{}'.format(kfp.__version__)
+    pipeline_spec.schema_version = 'v2alpha1'
+
+    # Pipeline Parameters
+    for arg in args:
+      if arg.value is not None:
+        if isinstance(arg.value, int):
+          pipeline_spec.runtime_parameters[
+              arg.name].type = pipeline_spec_pb2.PrimitiveType.INT
+          pipeline_spec.runtime_parameters[
+              arg.name].default_value.int_value = arg.value
+        elif isinstance(arg.value, float):
+          pipeline_spec.runtime_parameters[
+              arg.name].type = pipeline_spec_pb2.PrimitiveType.DOUBLE
+          pipeline_spec.runtime_parameters[
+              arg.name].default_value.double_value = arg.value
+        elif isinstance(arg.value, str):
+          pipeline_spec.runtime_parameters[
+              arg.name].type = pipeline_spec_pb2.PrimitiveType.STRING
+          pipeline_spec.runtime_parameters[
+              arg.name].default_value.string_value = arg.value
+        else:
+          raise NotImplementedError(
+              'Unexpected parameter type with: "{}".'.format(str(arg.value)))
+
+    deployment_config = pipeline_spec_pb2.PipelineDeploymentConfig()
+    importer_tasks = []
+
+    for op in pipeline.ops.values():
```

Comment:
You might want to sort the ops to prevent problems when using python 3.5:

```python
from collections import OrderedDict

for op in OrderedDict(sorted(pipeline.ops)).values():
```

Reply:
Thank you for helping debugging the test failure. Fixed via https://github.com/kubeflow/pipelines/pull/4529/files#diff-ba820c40ad57f55edd0d1eaab8ec53b2075893ef43640b73fcefffb5e0fe30a5R109